AK/BigIntBase.h

```diff
@@ -0,0 +1,566 @@
+#    endif
+        return;
+#else
+#    if defined(AK_COMPILER_CLANG)
+        NativeWord ncarry;
+        if constexpr (SameAs<NativeWord, unsigned int>)
+            output = __builtin_subc(word1, word2, carry, reinterpret_cast<unsigned int*>(&ncarry));
+        else if constexpr (SameAs<NativeWord, unsigned long>)
+            output = __builtin_subcl(word1, word2, carry, reinterpret_cast<unsigned long*>(&ncarry));
+        else if constexpr (SameAs<NativeWord, unsigned long long>)
+            output = __builtin_subcll(word1, word2, carry, reinterpret_cast<unsigned long long*>(&ncarry));
+        else
+            VERIFY_NOT_REACHED();
+        carry = ncarry;
+        return;
+#    endif
+#endif
+    }
+    bool ncarry = __builtin_sub_overflow(word1, word2, &output);
+    if (carry) {
+        if (output == 0)
+            ncarry = true;
+        --output;
+    }
+    carry = ncarry;
+}
+
+// Calculate ((divident1 << word_size) + divident0) / divisor. Quotient should be guaranteed to fit into NativeWord.
+ALWAYS_INLINE constexpr NativeWord div_mod_words(NativeWord divident0, NativeWord divident1, NativeWord divisor, NativeWord& remainder)
+{
+#if ARCH(x86_64)
+    if (!__builtin_is_constant_evaluated()) {
+        NativeWord remainder_value, quotient;
+        asm("div %3"
+            : "=a"(remainder_value), "=d"(quotient)
+            : "a"(divident0), "d"(divident1), "rm"(divisor));
+        remainder = remainder_value;
+        return quotient;
+    }
+#endif
+    auto divident = (static_cast<DoubleWord>(divident1) << word_size) + divident0;
+    remainder = static_cast<NativeWord>(divident % divisor);
+    return static_cast<NativeWord>(divident / divisor);
+}
+
+// ===== Operations on integer storages =====
+constexpr void storage_copy(auto const& operand, auto&& result, size_t offset = 0)
+{
+    auto fill = extend_sign(operand.is_negative());
+    size_t size1 = operand.size(), size = result.size();
+
+    for (size_t i = 0; i < size; ++i)
+        result[i] = i + offset < size1 ? operand[i + offset] : fill;
+}
+
+constexpr void storage_set(NativeWord value, auto&& result)
+{
+    result[0] = value;
+    for (size_t i = 1; i < result.size(); ++i)
+        result[i] = 0;
+}
+
+// `is_for_inequality' is a hint to compiler that we do not need to differentiate between < and >.
+constexpr int storage_compare(auto const& operand1, auto const& operand2, bool is_for_inequality)
+{
+    bool sign1 = operand1.is_negative(), sign2 = operand2.is_negative();
+    size_t size1 = operand1.size(), size2 = operand2.size();
+
+    if (sign1 != sign2) {
+        if (sign1)
+            return -1;
+        return 1;
+    }
+
+    NativeWord compare_value = extend_sign(sign1);
+    bool differ_in_high_bits = false;
+
+    if (size1 > size2) {
+        for (size_t i = size1; i-- > size2;)
+            if (operand1[i] != compare_value)
+                differ_in_high_bits = true;
```

Comment:
looks like you could `break` immediately after match (in both cases); you could `return` if youʼd decide to extract that to a(n inlined) function.

Reply:
That's true but (a) compiler might be already smart enough to do so; (b) see above.